cpp/src/arrow/compute/kernels/vector_selection.cc

```diff
@@ -2146,6 +2147,144 @@ class TakeMetaFunction : public MetaFunction {
   }
 };
 
+// ----------------------------------------------------------------------
+// DropNull Implementation
+
+Result<std::shared_ptr<arrow::Array>> GetNotNullIndices(
+    const std::shared_ptr<Array>& column, MemoryPool* memory_pool) {
+  std::shared_ptr<arrow::Array> indices;
+  arrow::NumericBuilder<arrow::Int32Type> builder(memory_pool);
```

Comment:
I think instead of explicitly building indices of valid bits, it'd be more compact and performant to reuse GetTakeIndices. You can zero-copy construct a BooleanArray which is true where `column` is valid and false where `column` is null, then pass it to GetTakeIndices to acquire `indices`

Reply:
If you had such a boolean array (and this may be future work) could it be more efficient to have a take that accepts a boolean array (of the same size as the input) and doesn't even construct the intermediate indices array?